### PR TITLE
[#8] Add an option to fold subtrees of already compared derivations

### PR DESCRIPTION
**Pull request by s-and-witch**
Problem: User may want to skip derivations, that has only already compared input below

Solution: Add new option to the DerivationDiff, add transformation, that will fold such subtrees, and add new cli option to enable it.

### CODE REVIEW

File: test/Main.hs [hunk 1]
```diff
@@ -3,11 +3,20 @@ module Main where
...
+main = do
+  quickCheck prop_DecodeEncodeID
+  quickCheck prop_foldingComesToFixedPoint
 
 prop_DecodeEncodeID :: DerivationDiff -> Bool
 prop_DecodeEncodeID dd = case decode (encode dd) of
   Nothing -> False
   Just dd' -> dd == dd'
+
+prop_foldingComesToFixedPoint :: DerivationDiff -> Bool
```
Comment (rvem) [inline test/Main.hs]: I don't think there is a high chance that `Arbitrary` will generate a value that will change after `foldAlreadyComparedSubTrees`. Could you please check if that actually happens during these property tests and if so how frequently. If it doesn't happen at least couple of times on each test run, I think we should add some unit tests

File: src/Nix/Diff/Types.hs [hunk 2]
```diff
@@ -94,8 +94,9 @@ type Argument = Text
 data DerivationDiff
   = DerivationsAreTheSame
   | AlreadyCompared
-  | NamesDontMatch   { outputStructure :: Changed OutputStructure}
-  | OutputsDontMatch { outputStructure :: Changed OutputStructure}
+  | OnlyAlreadComparedBelow { outputStructure :: Changed OutputStructure}
```
Comment (rvem) [inline src/Nix/Diff/Types.hs]:
```suggestion
  | OnlyAlreadyComparedBelow { outputStructure :: Changed OutputStructure}
```

File: src/Nix/Diff/Transformations.hs [hunk 3]
```diff
@@ -0,0 +1,87 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE MultiWayIf #-}
+module Nix.Diff.Transformations where
+
+import Nix.Diff.Types
+
+{-| In large diffs there may be a lot of derivations
```
Comment (rvem) [inline src/Nix/Diff/Transformations.hs:9]: Do you have an example of two such derivations when `--skip-already-compared` reduces the resulting output?

File: src/Nix/Diff/Render/HumanReadable.hs [hunk 4]
```diff
@@ -133,6 +133,9 @@ renderDiffHumanReadable :: DerivationDiff -> Render ()
 renderDiffHumanReadable = \case
     DerivationsAreTheSame -> pure ()
     AlreadyCompared ->  echo (explain "These two derivations have already been compared")
+    OnlyAlreadComparedBelow {..} -> do
+      renderOutputStructure outputStructure
+      echo (explain "Skipping as only already-compared derivations have changed below")
```
Comment (rvem) [inline src/Nix/Diff/Render/HumanReadable.hs]: Slightly more easy to understand wording
```suggestion
      echo (explain "Skipping because only derivations that have already been compared and shown in the diff are below")
```

File: test/Main.hs [hunk 5]
```diff
@@ -3,11 +3,20 @@ module Main where
...
+prop_foldingComesToFixedPoint dd = 
+   (op dd) == (op . op $ dd)
+  where
+    op = foldAlreadyComparedSubTrees
```
Comment (rvem) [inline test/Main.hs]: A small nitpick: newline at the end of file is missing. Same for `src/Nix/Diff/Transformations.hs`

**Review by rvem — COMMENTED** [review]
LGTM in general, but I would like to check tests a bit before merging

File: src/Nix/Diff/Transformations.hs (hunk 3 above)
Comment (s-and-witch) [reply src/Nix/Diff/Transformations.hs:9]: Yes, we can use example from readme for this purpose
```
29,31c29
<       • The input derivation named `dbus-1` differs
<         • These two derivations have already been compared
<       • Skipping environment comparison
---
>       • Skipping because only derivations that have already been compared and shown in the diff are below
35,37c33
<       • The input derivation named `system-path` differs
<         • These two derivations have already been compared
<       • Skipping environment comparison
---
>       • Skipping because only derivations that have already been compared and shown in the diff are below
41,43c37
<       • The input derivation named `system-path` differs
<         • These two derivations have already been compared
<       • Skipping environment comparison
---
>       • Skipping because only derivations that have already been compared and shown in the diff are below
80,86c74
<     • The input derivation named `unit-dbus.service` differs
<       - /nix/store/1h7np4jmaf2vc6sczvj7sg1hhinh2lzq-unit-dbus.service.drv:{out}
<       + /nix/store/llg852i3p19q17hw98gq11r61wbvjnrj-unit-dbus.service.drv:{out}
<       • The input derivation named `dbus-1` differs
<         • These two derivations have already been compared
<       • Skipping environment comparison
<     • Skipping environment comparison
---
>     • Skipping because only derivations that have already been compared and shown in the diff are below
```

File: test/Main.hs (hunk 1 above)
Comment (s-and-witch) [reply test/Main.hs]: There is a problem, that I don't know what corner cases are in this transformation. Let's just replace this test with new golden test

Comment (rvem) [reply test/Main.hs]: That works for me as well :)

**Review by rvem — APPROVED** [review]
